components/viaduct/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+    /// been set, set the Content-Type header to "application/json".
+    ///
+    /// Note: This panics if serde_json::to_vec fails. This can only happen
+    /// in a couple cases:
+    ///
+    /// 1. Trying to serialize a map with non-string keys.
+    /// 2. We wrote a custom serializer that fails.
+    ///
+    /// Neither of these are things we do. If they happen, it seems better for
+    /// this to fail hard with an easy to track down panic, than for e.g. `sync`
+    /// to fail with a JSON parse error (which we'd probably attribute to
+    /// corrupt data on the server, or something).
+    pub fn json<T: ?Sized + serde::Serialize>(mut self, val: &T) -> Self {
+        self.body =
+            Some(serde_json::to_vec(val).expect("Rust component bug: serde_json::to_vec failure"));
+        self.headers
+            .insert_if_missing(header_names::CONTENT_TYPE, "application/json")
+            .unwrap(); // We know this has to be valid.
+        self
+    }
+}
+
+/// A response from the server.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Response {
+    /// The method used to request this response.
+    pub request_method: Method,
+    /// The URL of this response.
+    pub url: Url,
+    /// The HTTP Status code of this response.
+    pub status: u16,
+    /// The headers returned with this response.
+    pub headers: Headers,
+    /// The body of the response. Note that responses with binary bodies are
+    /// currently unsupported.
+    pub body: Vec<u8>,
+}
+
+impl Response {
+    /// Parse the body as JSON.
+    pub fn json<'a, T>(&'a self) -> Result<T, serde_json::Error>
+    where
+        T: serde::Deserialize<'a>,
+    {
+        serde_json::from_slice(&self.body)
+    }
+
+    /// Get the body as a string. Assumes UTF-8 encoding. Any non-utf8 bytes
+    /// are replaced with the replacement character.
+    pub fn text(&self) -> std::borrow::Cow<str> {
+        String::from_utf8_lossy(&self.body)
+    }
+
+    /// Returns true if the status code is in the interval `[200, 300)`.
+    #[inline]
+    pub fn is_success(&self) -> bool {
+        status_codes::is_success_code(self.status)
+    }
+
+    /// Returns true if the status code is in the interval `[500, 600)`.
+    #[inline]
+    pub fn is_server_error(&self) -> bool {
+        status_codes::is_server_error_code(self.status)
+    }
+
+    /// Returns true if the status code is in the interval `[400, 500)`.
+    #[inline]
+    pub fn is_client_error(&self) -> bool {
+        status_codes::is_client_error_code(self.status)
+    }
+
+    /// Returns an [`UnexpectedStatus`] error if `self.is_success()` is false,
+    /// otherwise returns `Ok(self)`.
+    #[inline]
+    pub fn require_success(self) -> Result<Self, UnexpectedStatus> {
+        if self.is_success() {
+            Ok(self)
+        } else {
+            Err(UnexpectedStatus {
+                method: self.request_method,
+                // XXX We probably should try and sanitize this. Replace the user id
```

Comment:
I think it makes sense to do that in the callers, but I'm not sure if Viaduct is the right place to add that sanitization.

Reply:
A big, but fun, project might be to try to use the Rust typesystem to force such sanitization on all consumers via some sort of "taint checking" system. Not suggesting we do that with any urgency, but I do know that the alternative is a long tail of places where we accidentally log/expose sensitive data to places we didn't intend.